cmd/ilm-restore.go

```diff
@@ -0,0 +1,323 @@
+	CustomHelpTemplate: `NAME:
+  {{.HelpName}} - {{.Usage}}
+
+USAGE:
+  {{.HelpName}} TARGET
+
+DESCRIPTION:
+  Create a restored copy of one or more archived objects.
+
+EXAMPLES:
+  1. Restore one specific object
+     {{.Prompt}} {{.HelpName}} myminio/mybucket/path/to/object
+
+  2. Restore a specific object version
+     {{.Prompt}} {{.HelpName}} --vid "CL3sWgdSN2pNntSf6UnZAuh2kcu8E8si" myminio/mybucket/path/to/object
+
+  3. Restore all objects under a specific prefix
+     {{.Prompt}} {{.HelpName}} --recursive myminio/mybucket/dir/
+
+  4. Restore all objects with all versions under a specific prefix
+     {{.Prompt}} {{.HelpName}} --recursive --versions myminio/mybucket/dir/
+
+`,
+}
+
+// checkILMRestoreSyntax - validate arguments passed by user
+func checkILMRestoreSyntax(ctx *cli.Context) {
+	if len(ctx.Args()) != 1 {
+		cli.ShowCommandHelpAndExit(ctx, "restore", globalErrorExitStatus)
+	}
+
+	if ctx.Int("days") <= 0 {
+		fatalIf(errDummy().Trace(), "--days should be equal or greater than 1")
+	}
+
+	if ctx.Bool("version-id") && (ctx.Bool("recursive") || ctx.Bool("versions")) {
+		fatalIf(errDummy().Trace(), "You cannot combine --version-id with --recursive or --versions flags.")
+	}
+}
+
+// Send Restore S3 API
+func restoreObject(ctx context.Context, targetAlias, targetURL, versionID string, days int) *probe.Error {
+	clnt, err := newClientFromAlias(targetAlias, targetURL)
+	if err != nil {
+		return err
+	}
+
+	return clnt.Restore(ctx, versionID, days)
+}
+
+// Send restore S3 API request to one or more objects depending on the arguments
+func sendRestoreRequests(ctx context.Context, targetAlias, targetURL, targetVersionID string, recursive, applyOnVersions bool, days int, restoreSentReq chan *probe.Error) {
+	defer close(restoreSentReq)
+
+	client, err := newClientFromAlias(targetAlias, targetURL)
+	if err != nil {
+		restoreSentReq <- err
+		return
+	}
+
+	if !recursive {
+		err := restoreObject(ctx, targetAlias, targetURL, targetVersionID, days)
+		restoreSentReq <- err
+		return
+	}
+
+	prev := ""
+	for content := range client.List(ctx, ListOptions{
+		Recursive:         true,
+		WithOlderVersions: applyOnVersions,
+		ShowDir:           DirNone,
+	}) {
+		if content.Err != nil {
+			errorIf(content.Err.Trace(client.GetURL().String()), "Unable to list folder.")
+			continue
+		}
+		err := restoreObject(ctx, targetAlias, content.URL.String(), content.VersionID, days)
+		if err != nil {
+			restoreSentReq <- err
+			continue
+		}
+		// Avoid sending the status of each separate version
+		// of the same object name.
+		if prev != content.URL.String() {
+			prev = content.URL.String()
+			restoreSentReq <- nil
+		}
+	}
+}
+
+// Wait until an object which receives restore request is completely restored in the fast tier
+func waitRestoreObject(ctx context.Context, targetAlias, targetURL, versionID string) *probe.Error {
+	clnt, err := newClientFromAlias(targetAlias, targetURL)
+	if err != nil {
+		return err
+	}
+
+	for {
```

Comment:
shouldn't we wait between successive Stat requests to fetch restore status?

Reply:
> shouldn't we wait between successive Stat requests to fetch restore status?

right, fixed